README.md
# node-efsw

Watching your filesystem events just like [fsevents](https://www.npmjs.com/package/fsevents). But **node-efsw** is cross-platform.

> Node.js binding for [efsw](https://github.com/SpartanJ/efsw).

## Installation

```sh
$ npm install --save efsw
```

## Usage

```js
const Watcher = require("efsw").Watcher;

const watcher = new Watcher("/tmp");
watcher.on("change", function(filename, info) {
    console.log(filename, info);
});
watcher.on("error", function(err) {
    console.error(err);
});

watcher.start();
watcher.stop();
```

## API

### Constructor

```js
new Watcher(dir);
```

#### Parameters

+ `dir`: the directory to be watched.

### `start()`

Start the watcher. It may occur some error events, eg. the directory is not exist.

### `stop()`

Stop the watcher.

### Events

#### `.on("change", function(filename, info) {})`

This event will be emitted when a filesystem event occurred, eg. a file added.

##### Callback Parameters

+ `filename`: the absolute filename;
+ `info`: the event information object;
    - `info.dir`: the base directory that this watcher is watching;
    - `info.action`: the event type:
        * `"ADD"`
        * `"DELETE"`
        * `"MODIFIED"`
        * `"MOVED"`
    - `info.relative`: the relative filename that based from `info.dir`;
    - `info.old`: if the event is `"MOVED"`, this column indicates the old filename;
    - `info.oldRelative`: similar with `info.relative`. If the event is `"MOVED"`, this column indicates the old filename.

#### `.on("error", function(err) {})`

This event will be emitted when an error occurred.

##### Callback Parameters

+ `err`: the error object.

## Contribution

You're welcome to make pull request.


History.md

1.0.4 / 2021-03-30
==================

**others**
  * [[`78d350f`](http://github.com/XadillaX/node-efsw/commit/78d350f1c80c5d0fc234505ae620905c36636e1d)] - chore: update efsw (XadillaX <<i@2333.moe>>)

1.0.3 / 2020-02-11
==================

**others**
  * [[`f300622`](http://github.com/XadillaX/node-efsw/commit/f300622596db12632e6ccac914f9ab97ccad4f9e)] - chore: update efsw (XadillaX <<i@2333.moe>>)

1.0.2 / 2020-02-07
==================

**others**
  * [[`ff59bf5`](http://github.com/XadillaX/node-efsw/commit/ff59bf5178279dcd8374f47488fd13d30c618a8f)] - chore: update efsw (XadilaX <<i@2333.moe>>)

1.0.1 / 2020-01-19
==================

**others**
  * [[`92d680f`](http://github.com/XadillaX/node-efsw/commit/92d680f1b4116aa0b7b0070b188f10ad95e5737a)] - docs: update LICENSE (XadillaX <<i@2333.moe>>),fatal: No names found, cannot describe anything.



src/efsw_core.h
/**
 * node-efsw - Node.js binding for EFSW
 *
 * Copyright (c) 2017 XadillaX <i@2333.moe>
 *
 * MIT License <https://github.com/XadillaX/node-efsw/blob/master/LICENSE>
 */
#ifndef __EFSW_CORE_H__
#define __EFSW_CORE_H__
#include <string>
#include <nan.h>
#include "./deps/efsw/include/efsw/efsw.hpp"
#include "efsw_core_listener.h"

namespace efsw_core {

class EFSWCore : public Nan::ObjectWrap {
public:
    static NAN_MODULE_INIT(Init);

private:
    explicit EFSWCore(const char* path, Nan::Callback* listener);
    ~EFSWCore();

    efsw::WatchId Start();
    void Stop();

    static NAN_METHOD(New);
    static NAN_METHOD(Start);
    static NAN_METHOD(Stop);

private:
    std::string path;
    Nan::Callback* listener;
    efsw::FileWatcher* watcher;
    efsw::WatchId watch_id;
    EFSWCoreListener core_listener;
};

}

#endif


src/efsw_core.cc
/**
 * node-efsw - Node.js binding for EFSW
 *
 * Copyright (c) 2017 XadillaX <i@2333.moe>
 *
 * MIT License <https://github.com/XadillaX/node-efsw/blob/master/LICENSE>
 */
#include "efsw_core.h"
using namespace efsw;
using namespace v8;

namespace efsw_core {

EFSWCore::EFSWCore(const char* path, Nan::Callback* listener) :
    path(path),
    listener(listener),
    core_listener(listener)
{
    watcher = new FileWatcher();
    watch_id = 0;
}

EFSWCore::~EFSWCore()
{
    Nan::HandleScope scope;
    delete watcher;
    delete listener;
}

WatchId EFSWCore::Start()
{
    if(watch_id > 0)
    {
        return -100;
    }

    WatchId ret = watcher->AddWatch(path.c_str(), &core_listener, true);
    if(ret > 0) {
        watch_id = ret;
        watcher->watch();
    }

    return ret;
}

void EFSWCore::Stop()
{
    if(watch_id <= 0) return;

    watcher->RemoveWatch(watch_id);
    watch_id = 0;
}

NAN_METHOD(EFSWCore::New)
{
    v8::Isolate* isolate = info.GetIsolate();
    Local<Context> context = isolate->GetCurrentContext();

    Nan::Utf8String path(info[0]->ToString(context).ToLocalChecked());
    Nan::Callback* listener = new Nan::Callback(info[1].As<v8::Function>());

    EFSWCore* core = new EFSWCore(*path, listener);
    core->Wrap(info.This());
    info.GetReturnValue().Set(info.This());
}

NAN_METHOD(EFSWCore::Start)
{
    EFSWCore* core = Nan::ObjectWrap::Unwrap<EFSWCore>(info.Holder());
    WatchId watch_id = core->Start();

    if(watch_id < 0)
    {
        if(watch_id == -100)
        {
            return Nan::ThrowError("Watcher is already in watching.");
        }

        return Nan::ThrowError(Errors::Log::GetLastErrorLog().c_str());
    }
}

NAN_METHOD(EFSWCore::Stop)
{
    EFSWCore* core = Nan::ObjectWrap::Unwrap<EFSWCore>(info.Holder());
    core->Stop();
}

NAN_MODULE_INIT(EFSWCore::Init)
{
    v8::Local<v8::FunctionTemplate> tpl = Nan::New<v8::FunctionTemplate>(New);
    tpl->SetClassName(Nan::New("EFSWCore").ToLocalChecked());
    tpl->InstanceTemplate()->SetInternalFieldCount(1);

    Nan::SetPrototypeMethod(tpl, "start", Start);
    Nan::SetPrototypeMethod(tpl, "stop", Stop);

    Nan::Set(target, Nan::New("EFSWCore").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
}

}


src/efsw_core_listener.h
/**
 * node-efsw - Node.js binding for EFSW
 *
 * Copyright (c) 2017 XadillaX <i@2333.moe>
 *
 * MIT License <https://github.com/XadillaX/node-efsw/blob/master/LICENSE>
 */
#ifndef __EFSW_CORE_LISTENER_H__
#define __EFSW_CORE_LISTENER_H__
#include <string>
#include <nan.h>
#include "./deps/efsw/include/efsw/efsw.hpp"

namespace efsw_core {

#define Watch watch
#define AddWatch addWatch
#define RemoveWatch removeWatch
#define HandleFileAction handleFileAction
#define GetLastErrorLog getLastErrorLog
#define WatchId WatchID

class EFSWCoreListener : public efsw::FileWatchListener {
public:
    EFSWCoreListener(Nan::Callback* listener);
    ~EFSWCoreListener();

    void HandleFileAction(
            efsw::WatchId,
            const std::string& dir,
            const std::string& filename,
            efsw::Action action,
            std::string old_filename = "");

private:
    Nan::Callback* listener;
};

}

#endif


src/efsw_core_listener.cc
/**
 * node-efsw - Node.js binding for EFSW
 *
 * Copyright (c) 2017 XadillaX <i@2333.moe>
 *
 * MIT License <https://github.com/XadillaX/node-efsw/blob/master/LICENSE>
 */
#include "efsw_core_listener.h"
using namespace std;
using namespace efsw;
using namespace v8;

namespace efsw_core {

struct AsyncArgs {
    string dir;
    string filename;
    string old_filename;
    Action action;
    Nan::Callback* callback;

    uv_async_t handle;
};

void OnAsyncClosed(uv_handle_t* handle)
{
    uv_async_t* async = (uv_async_t*)handle;
    AsyncArgs* data = (AsyncArgs*)async->data;
    delete data;
}

void OnEFSWEvent(uv_async_t* handle)
{
    Nan::HandleScope scope;

    Nan::AsyncResource resource("efsw:core_listener.on_efsw_event");

    AsyncArgs* data = (AsyncArgs*)handle->data;
    Local<Value> argv[] = {
        Nan::New(data->dir.c_str()).ToLocalChecked(),
        Nan::New(data->filename.c_str()).ToLocalChecked(),
        Nan::New(data->old_filename.c_str()).ToLocalChecked(),
        Nan::New(data->action)
    };
    data->callback->Call(4, argv, &resource);
    uv_close((uv_handle_t*)handle, OnAsyncClosed);
}

EFSWCoreListener::EFSWCoreListener(Nan::Callback* listener) :
    listener(listener)
{
}

EFSWCoreListener::~EFSWCoreListener()
{
}

void EFSWCoreListener::HandleFileAction(
        efsw::WatchId,
        const std::string& dir,
        const std::string& filename,
        efsw::Action action,
        std::string old_filename)
{
    AsyncArgs* args = new AsyncArgs();
    args->dir = dir;
    args->filename = filename;
    args->old_filename = old_filename;
    args->action = action;
    args->callback = listener;

    uv_async_init(uv_default_loop(), &args->handle, OnEFSWEvent);
    args->handle.data = (void*)args;
    uv_async_send(&args->handle);
}

}


lib/watcher.js
/**
 * XadillaX <i@2333.moe> created at 2017-08-22 16:19:31 with ❤
 *
 * Copyright (c) 2017 xcoder.in, all rights reserved.
 */
"use strict";

const EventEmitter = require("events").EventEmitter;
const fs = require("fs");
const path = require("path");

const EFSWCore = require("../build/Release/core").EFSWCore;

const ACTION_MAP = {
    "1": "ADD",
    "2": "DELETE",
    "3": "MODIFIED",
    "4": "MOVED"
};

class Watcher extends EventEmitter {
    constructor(_path) {
        _path = path.resolve(process.cwd(), _path);

        super();
        Object.defineProperties(this, {
            path: {
                value: _path,
                configurable: false,
                enumerable: true,
                writable: false
            },
            core: {
                value: new EFSWCore(_path, this._onEFSWEvent.bind(this)),
                configurable: false,
                enumerable: false,
                writable: false
            }
        });
        this.watching = false;
    }

    start() {
        if(this.watching) {
            return this.emit("error", new Error("Watcher is already in watching."));
        }

        try {
            this.core.start();
        } catch(e) {
            return this.emit("error", e);
        }
        this.watching = true;
    }

    stop() {
        this.core.stop();
        this.watching = false;
    }

    _onEFSWEvent(_path, filename, oldFilename, action) {
        const fullPath = path.join(_path, filename);
        const info = {
            dir: _path,
            action: ACTION_MAP[action] || "UNKNOWN",
            relative: filename
        };

        if(oldFilename) {
            info.old = path.join(_path, oldFilename);
            info.oldRelative = oldFilename;
        }

        if(info.action === "UNKNOWN") {
            const err = new Error(`Unknown fs event ${action}.`);
            err.info = info;
            this.emit("error", err);
        } else {
            this.emit("change", fullPath, info);
        }
    }
}

module.exports = Watcher;
